runtime/vm/resolvesupport.cpp

```diff
@@ -1839,6 +1839,93 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
 
 }
 
+j9object_t
+resolveConstantDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpIndex, UDATA resolveFlags)
+{
+	J9RAMConstantDynamicRef *ramCPEntry = (J9RAMConstantDynamicRef*)ramCP + cpIndex;
+	j9object_t value = ramCPEntry->value;
+
+	J9ROMClass *romClass = ramCP->ramClass->romClass;
+	J9ROMConstantDynamicRef *romConstantRef = (J9ROMConstantDynamicRef*)(J9_ROM_CP_FROM_CP(ramCP) + cpIndex);
+	J9SRP *callSiteData = (J9SRP *) J9ROMCLASS_CALLSITEDATA(romClass);
+	U_16 *bsmIndices = (U_16 *) (callSiteData + romClass->callSiteCount);
+	U_16 *bsmData = bsmIndices + romClass->callSiteCount;
+	J9ROMNameAndSignature* nameAndSig = SRP_PTR_GET(&romConstantRef->nameAndSignature, J9ROMNameAndSignature*);
+	U_32 bsmIndex = romConstantRef->bsmIndexAndCpType >> J9DescriptionCpTypeShift;
+	J9MemoryManagerFunctions *gcFuncs = vmThread->javaVM->memoryManagerFunctions;
+	UDATA resolved = 0;
+
+	/* Check if already resolved */
+	omrthread_monitor_enter(vmThread->javaVM->constantDynamicMutex);
+	if (NULL != value) {
+		resolved = 1;
+	} else if (NULL != ramCPEntry->exception) {
+		VM_VMHelpers::setExceptionPending(vmThread, ramCPEntry->exception);
+		resolved = 1;
+	}
+	omrthread_monitor_exit(vmThread->javaVM->constantDynamicMutex);
+
+	/* Enter if not previously resolved */
+	if (0 == resolved) {
+		/* Walk bsmData */
+		for (U_32 i = 0; i < bsmIndex; i++) {
+			bsmData += bsmData[1] + 2;
+		}
+
+		sendResolveConstantDynamic(vmThread, ramCP, cpIndex, nameAndSig, bsmData);
+		value = (j9object_t)vmThread->returnValue;
+
+		/* lock write access to constantDynamic ram slots */
+		omrthread_monitor_enter(vmThread->javaVM->constantDynamicMutex);
```

Comment:
Similar concern here.  We need to spend some time thinking about the ordering, memory barriers, and CAS to handle setting and reading these fields without a lock